contracts/Committee.sol

```diff
@@ -5,79 +5,224 @@ import "./spec_interfaces/IValidatorsRegistration.sol";
+	function findFreePos(CommitteeInfo memory info) private pure returns (uint8 pos) {
+		pos = 0;
+		uint32 bitmap = info.committeeBitmap;
+		while (bitmap & 1 == 1) {
+			bitmap >>= 1;
+			pos++;
+		}
+	}
+
+	function qualifiesToEnterCommittee(address addr, MemberData memory data, CommitteeInfo memory info, Settings memory _settings) private view returns (bool qualified, uint8 entryPos) {
+		if (!data.isMember || data.weight == 0) {
+			return (false, 0);
+		}
+
+		if (info.committeeSize < _settings.maxCommitteeSize) {
+			return (true, findFreePos(info));
+		}
+
+		CommitteeMember memory minMember = committee[info.minCommitteeMemberPos];
+		if (data.weight < minMember.weight || data.weight == minMember.weight && addr < minMember.addr) {
+			return (false, 0);
+		}
+
+		return (true, info.minCommitteeMemberPos);
+	}
+
+	function saveMemberData(address addr, MemberData memory data) private {
+		if (data.isMember) {
+			membersData[addr] = data;
+		} else {
+			delete membersData[addr];
+		}
+	}
+
+	function updateOnMemberChange(address addr, MemberData memory data) private returns (bool committeeChanged) {
+		CommitteeInfo memory info = committeeInfo;
+		Settings memory _settings = settings;
+		bytes memory sortBytes = abi.encodePacked(committeeSortBytes);
+
+		if (!data.inCommittee) {
+			(bool qualified, uint8 entryPos) = qualifiesToEnterCommittee(addr, data, info, _settings);
+			if (!qualified) {
+				saveMemberData(addr, data);
+				return false;
+			}
+
+			(info, sortBytes) = removeMemberAtPos(entryPos, sortBytes, info);
+			(info, sortBytes) = addToCommittee(addr, data, entryPos, sortBytes, info);
+		}
+
+		(info, sortBytes) = (data.isMember && data.weight > 0) ?
+			rankMember(addr, data, sortBytes, info) :
+			removeMemberFromCommittee(data, sortBytes, info);
+
+		emit ValidatorCommitteeChange(addr, data.weight, data.isCompliant, data.inCommittee);
+
+		saveMemberData(addr, data);
+
+		committeeInfo = info;
+		committeeSortBytes = sortBytes.toBytes32(0);
+
+		assignRewardsIfNeeded(_settings);
+
+		return true;
+	}
+
+	function addToCommittee(address addr, MemberData memory data, uint8 entryPos, bytes memory sortBytes, CommitteeInfo memory info) private returns (CommitteeInfo memory newInfo, bytes memory newSortByes) {
+		committee[entryPos] = CommitteeMember({
+			addr: addr,
+			weight: data.weight
+		});
+		data.inCommittee = true;
+		data.pos = entryPos;
+		info.committeeBitmap |= uint32(uint(1) << entryPos);
+		info.committeeSize++;
+
+		sortBytes[info.committeeSize - 1] = byte(entryPos);
+		return (info, sortBytes);
+	}
+
+	function removeMemberFromCommittee(MemberData memory data, bytes memory sortBytes, CommitteeInfo memory info) private returns (CommitteeInfo memory newInfo, bytes memory newSortBytes) {
+		uint rank = 0;
+		while (uint8(sortBytes[rank]) != data.pos) {
+			rank++;
+		}
+
+		for (; rank < info.committeeSize - 1; rank++) {
+			sortBytes[rank] = sortBytes[rank + 1];
+		}
+		sortBytes[rank] = 0;
```

Comment:
info.committeeSize--;
sortBytes[info.committeeSize] = 0 
so it's easier to understand?